### ISSUE TITLE
Mismatch between parsing of pipes in nested wikitext.

### ISSUE THREAD

**Issue by kristian-clausal** [tatuylonen/wiktextract#208]
[Example word: αγανακτώ/Greek](https://en.wiktionary.org/wiki/%CE%B1%CE%B3%CE%B1%CE%BD%CE%B1%CE%BA%CF%84%CF%8E#Greek)
(Greek in generals has a ton of problems like these)

The output of [el-verb](https://en.wiktionary.org/wiki/Template:el-verb) template in the article is the sensible
> αγανακτώ • (aganaktó) (past αγανάκτησα, passive —, ppp αγανακτισμένος)

but for that last part, ppp, we get 
```json
    {
      "form": "[[participle",
      "tags": [
        "passive"
      ]
    },
```
which comes straight from the el-verb template:
```
-->{{#if:{{{ppp|}}}|, ''[[participle|<span title="passive perfect participle">ppp</span>]]'' '''[[{{{ppp|}}}]]'''}}{{#if:{{{ppp2|}}}| /'''[[{{{ppp2|}}}]]'''}}{{#if:{{{ppp3|}}}| /'''[[{{{ppp3|}}}]]'''}}<!--
```
The issue is not:
  * the "-" parameter of the passive in the template arguments
  * the italics surrounding the [[participle|...]]

but something a bit more annoying and fundamental: the "|" pipe symbol right after `[[participle|` is being parsed by us as being part of the larger {{#if:...}} block, and thus everything that comes after it is considered to be the "else" parameter of it all. Come to think of it, that explains all the "ppp"s I saw in the debug files, compare our example with [αφοδεύω](https://en.wiktionary.org/wiki/%CE%B1%CF%86%CE%BF%CE%B4%CE%B5%CF%8D%CF%89) where we have a field looking like
```json
    {
      "form": "1ppp]] [[]]",
      "tags": [
        "passive"
      ]
    },
```
because it does not have a `ppp` parameter.

The annoying thing is, I believe using a naked pipe inside template arguments like that is actually not kosher in wikitext; you have the "template" [{{!}}](https://en.wikipedia.org/wiki/Template:!) for that, a way to insert pipe characters to be evaluated later. And indeed, changing the pipes after `[[participle|` into `[[participle{{!}}` fixes the issue on our end, and probably doesn't break anything if applied to the wiktionary template.

IMO, this looks like it's an error-recovery mechanism in wikitext (or maybe just better nesting parsing) that corrects a mistake that we do not.

I'm going to ask Tatu to take a quick look at this next week because I haven't yet delved into the wikitext parsing stuff that much, and hopefully he can figure out the correct way to do this.

**Comment by kristian-clausal** [comment]
Seems I was wrong about {{!}}, [according to this.](https://meta.wikimedia.org/wiki/Help:Template#A_parameter_value_containing_a_pipe_character) If a pipe-character (inside a template argument) is part of `a template call, parser function call, piped link or image tag,` then you should use {{!}}, although so tables mostly. I don't think I've come across an example where a template argument splits one of those (template/parserfn/link/img) into component pieces like a table commonly is, so I guess that makes sense.

**Comment by xxyzz** [comment]
That vertical bar is for [renamed link](https://en.wikipedia.org/wiki/Help:Wikitext#Renamed_link).

**Comment by kristian-clausal** [comment]
I think I figured out what the problem is, at least:

```
-->{{#if:{{{ppp|}}}|;; BBB, [[foo|<span title="passive perfect participle">ppp</span>]] '''[[{{{ppp|}}}]]'''}}<!--
-->{{#if:{{{ppp|}}}|;; BB2, {{abbr|TEST1|test1}} '''[[{{{ppp|}}}]]'''}}<!--
-->{{#if:{{{ppp|}}}|;; BB3, [[borb|brob]] '''[[{{{ppp|}}}]]'''}}<!--
```
Three different versions of the offending #if block, with the outputs:

```
if_fn: ('αγανακτισμένος', ';; BBB, [[foo', '<span title="passive perfect participle">ppp</span>]] \'\'\'[[αγανακτισμένος]]\'\'\'')
...
if_fn: ('αγανακτισμένος', ";; BB2, \U00102089 '''[[αγανακτισμένος]]'''")
...
if_fn: ('αγανακτισμένος', ";; BB3, [[borb|brob]] '''[[αγανακτισμένος]]'''")
```
The lost of arguments inside each #if-block, as interpreted by our parser; the first is the broken one, and it seems like the HTML-code inside of the link argument is what's causing things to break.
In the second example (`{{abbr}}`) the abbr-template has been turned into a magic unicode character (so that it can be ignored when expanding templates outside of it and then reinserted by replacing the character with the original text).

Although in the JSON we find this...
```
    {
      "args": {
        "passive": "bar",
        "past": "αγανάκτησα",
        "ppp": "αγανακτισμένος"
      },
      "expansion": "αγανακτώ • (aganaktó)(past αγανάκτησα, passive bar;; BB
B, foo;; BB2, TEST1 [[αγανακτισμένος;; BB3, brob αγανακτισμένος)",
      "name": "el-verb"
    }
```

So the line with the {{abbr}}-template still has some artifacts left in the form of the opening brackets ("[[") of the following link.

**Comment by kristian-clausal** [comment]
Fixed in wikitextprocessor ab3e0dffac8e364747b0a56af6133eb100b373e7, but might be some other issues down the line, as usual.

Turns out it was a very simple two-character fix in a regex. When parsing, wikitextprocessor encodes whole templates and other node-generating stuff into magical characters so that they can be ignored while the stuff outside of them is being parsed, and then later they're uncompressed and processed then. The regex that did this for LINKs ([[links]]) specifically had a bit that forbid HTML tags, which isn't actually forbidden in wikitext, so when it encountered an example of this it wouldn't compress the link and this broke it up, creating garbage text or garbage parameters for outer templates, like what happened with the aganaktos example in this thread.

Tatu wasn't sure why he'd put the ``^><`` in the regex, but it's possible it was a fix for some other issue that will now pop up.
